docs/proposals/agent-healthcheck.md

```diff
@@ -0,0 +1,68 @@
+
+"Business" errors that would not benefit from restarting the agent (eg. integration warnings, configuration parsing errors, internet connectivity issues, ...) should not trigger an unhealthy status, but be signaled through the existing systems. 
+
+## Constraints
+
+- The list of components to monitor cannot be hardcoded, as some can be disabled by the configuration
+- The healthcheck query should be as fast as possible to avoid timeout-ing
+- The runtime cost of the query should be as low as possible
+- It should be generic enough to support both Docker and Systemd, and future needs
+
+## Recommended Solution
+
+### Have the components refresh their health status in a pluggable healthz registry
+
+This registry would be self-contained in it's own `healthz` package and expose the following API:
+
+- registration and de-registration of components as monitorable
+- ping from components to signal they are still healthy (after a configurable timeout with no ping, they will be considered unhealthy)
+- query interface to retrieve the health state: for every registered component, compare last ping timestamp and timeout to determine health status
+
+We will be able to address the two following use cases:
+
+- for the docker healthcheck, we will add an endpoint in the agent API that will query the health state. This will be achieved via the bundled `curl` binary
+- for the systemd integration, we will add a `healthz/systemd` subpackage that spawns a goroutine querying the state and reporting to the systemd API
+
+On a typical agent, we should monitor the following components:
+
+- dogstatsd intake loop + parsing loop (if either UDP or UDS is configured)
+- aggregator
+- check scheduling goroutine
+- autodiscovery, if a listener is registered
+- tagger, if it manages to init at least one collector
+- forwarder
+
+As a first approach, we'll make them register at the end of their init phase, then create a ticker that calls the health update from their main goroutine(s). The logic can then be improved component by component, depending on what "healthy" means for it.
+
+## Other solutions
+
+### Extend the `pkg/status` package logic
+
+The `pkg/status` package defines an standard way for components to expose statistics to be displayed on the `agent status` page. It is backed by `goexpvar`. That could be achieved by updating an `int64` timestamp as a goexpvar value, and parsing it.
+
+The drawback is that we will be "polluting" the expvar namespace with these timestamps. Plus, we don't have an obvious way to do component registration at runtime, so we'll need to introduce a registry for that.
+
+## Open Questions
+
+- As we already have the integration warning system and the `status` / `diagnose` commands, the health could be reported as a simple healthy/unhealthy boolean state. We could also extend the available states to a third "erroring-out, user action required, but restarting will not fix it" state (apikey errors, connectivity issues...).
```

Comment:
Let's start with a boolean approach (OK, FAIL). It is well supported across systems and orchestrators, and easy to parse programatically. More complex responses make sense in the status command as they are supposed to be consumed by humans.

Reply:
I may prefer a false positive behavior than false negative.
Let say we are running the agent in a Kubernetes Pod, with probes and a bad **DD_API_KEY**, then the kubelet will report an unhealthy state and this will leads to a restart of this Pod.

Effectively a restart will not help but letting the agent running as healthy when it cannot really do its job it's not convenient from an Ops point of view.
I may prefer having those agents failing in crashLoopBackOff (purposefully designed for that kind of use cases) so it's obvious that we have a problem.
Otherwise I would need to parse the logs of a `1/1 Running Pod`, that is not properly running.

Reply:
Ok so I _think_ everyone here agrees that we should start with a boolean approach (OK, FAIL), but then we can disagree about which category each error belongs in. (unless I'm missing something?)

To me an Agent configured with one api key should be regarded as unhealthy when that api key is invalid. But less clear-cut cases (connectivity issues for instance) should make the agent still report itself as healthy (unless the Agent is sure the issue will never resolve on its own).

Reply:
indeed, an invalid apikey makes dsd unable to fill it's role as a custom metrics submitter, so it should fail the liveliness probe. Removing this open question and defining a healthy forwarder as `still healthy on http errors, but unhealthy on invalid apikey`